### ISSUE TITLE
EurekaClient bean can never be instance of DiscoveryClient when @ConditionalOnRefreshScope matched

### ISSUE THREAD

**Issue by jearton** [spring-cloud/spring-cloud-netflix#2532]
# Version:
spring-boot:1.5.9.RELEASE  spring-cloud:Edgware.RELEASE

# Detail Problem:
In the method `EurekaHealthIndicator#getStatus`, I found the `if block` can never be entered because `eurekaClient instanceof DiscoveryClient` is always false. Follow is the source code:

```java
private Status getStatus(Builder builder) {
  Status status = new Status(
      this.eurekaClient.getInstanceRemoteStatus().toString(),
      "Remote status from Eureka server");

  if (eurekaClient instanceof DiscoveryClient && clientConfig.shouldFetchRegistry()) {
    DiscoveryClient discoveryClient = (DiscoveryClient) eurekaClient;
    long lastFetch = discoveryClient.getLastSuccessfulRegistryFetchTimePeriod();

    if (lastFetch < 0) {
      status = new Status("UP",
          "Eureka discovery client has not yet successfully connected to a Eureka server");
    }
    else if (lastFetch > clientConfig.getRegistryFetchIntervalSeconds() * 2000) {
      status = new Status("UP",
          "Eureka discovery client is reporting failures to connect to a Eureka server");
      builder.withDetail("renewalPeriod",
          instanceConfig.getLeaseRenewalIntervalInSeconds());
      builder.withDetail("failCount",
          lastFetch / clientConfig.getRegistryFetchIntervalSeconds());
    }
  }

  return status;
}
```

Then I figured out through these code in `EurekaClientAutoConfiguration$RefreshableEurekaClientConfiguration`:

```java
@Bean(destroyMethod = "shutdown")
@ConditionalOnMissingBean(value = EurekaClient.class, search = SearchStrategy.CURRENT)
@org.springframework.cloud.context.config.annotation.RefreshScope
@Lazy
public EurekaClient eurekaClient(ApplicationInfoManager manager, EurekaClientConfig config, EurekaInstanceConfig instance) {
  manager.getInfo(); // force initialization
  return new CloudEurekaClient(manager, config, this.optionalArgs,
      this.context);
}
```

Because of `@RefreshScope`,  the real type of `EurekaClient` bean is `com.sun.proxy.$ProxySth`.

My suggestion is to modify the method return type from `EurekaClient` to `CloudEurekaClient`, and it will be fine in `EurekaHealthIndicator`.

**Comment by spencergibb** [comment]
> My suggestion is to modify the method return type from EurekaClient to CloudEurekaClient, and it will be fine in EurekaHealthIndicator.

Not sure that will be the right fix, there is a spring utility to check if an instance is a proxy and check its type. But that is an implementation detail.

**Comment by jearton** [comment]
Yeah, I would prefer to use the spring utility.

**Comment by spencergibb** [comment]
Closed via 42449ee34e831edb31616c779ef4d8854078bd6a
